File: app/src/main/java/za/co/woolworths/financial/services/android/ui/adapters/SearchResultShopAdapter.java
package za.co.woolworths.financial.services.android.ui.adapters;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.text.Html;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.awfs.coordination.R;
import com.daimajia.swipe.SwipeLayout;
import com.daimajia.swipe.adapters.RecyclerSwipeAdapter;
import java.util.List;

import za.co.woolworths.financial.services.android.models.dto.OtherSkus;
import za.co.woolworths.financial.services.android.models.dto.ProductList;
import za.co.woolworths.financial.services.android.ui.adapters.holder.ProductListingViewType;
import za.co.woolworths.financial.services.android.ui.adapters.holder.SearchResultPriceItem;
import za.co.woolworths.financial.services.android.ui.fragments.shoppinglist.search.SearchResultNavigator;
import za.co.woolworths.financial.services.android.ui.views.WTextView;
import za.co.woolworths.financial.services.android.ui.views.WrapContentDraweeView;

public class SearchResultShopAdapter extends RecyclerSwipeAdapter<RecyclerView.ViewHolder> {

	private boolean value;

	private List<ProductList> mProductList;

	private SearchResultNavigator mSearchResultNavigator;

	private static final String FOOD_PRODUCT = "foodProducts";

	public SearchResultShopAdapter(List<ProductList> mProductList,
								   SearchResultNavigator searchResultNavigator) {
		this.mProductList = mProductList;
		this.mSearchResultNavigator = searchResultNavigator;
	}

	@Override
	public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		RecyclerView.ViewHolder vh;
		switch (ProductListingViewType.values()[viewType]) {
			case HEADER:
				vh = getHeaderViewHolder(parent);
				break;

			case FOOTER:
				vh = getProgressViewHolder(parent);
				break;

			case PRODUCT:
			default:
				vh = getSimpleViewHolder(parent);
				break;
		}
		return vh;
	}

	@NonNull
	private ProgressViewHolder getProgressViewHolder(ViewGroup parent) {
		return new ProgressViewHolder(LayoutInflater.from(parent.getContext()).inflate(
				R.layout.bottom_progress_bar, parent, false));
	}

	@NonNull
	private SimpleViewHolder getSimpleViewHolder(ViewGroup parent) {
		return new SimpleViewHolder(LayoutInflater.from(parent.getContext())
				.inflate(R.layout.shop_search_product_item, parent, false));
	}

	@NonNull
	private HeaderViewHolder getHeaderViewHolder(ViewGroup parent) {
		return new HeaderViewHolder(LayoutInflater.from(parent.getContext()).inflate(
				R.layout.item_found_layout, parent, false));
	}

	@Override
	public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
		final ProductList productList = mProductList.get(position);
		if (productList.rowType == ProductListingViewType.HEADER) {
			HeaderViewHolder hvh = (HeaderViewHolder) holder;
			hvh.setTotalItem(productList);
		} else if (productList.rowType == ProductListingViewType.FOOTER) {
			ProgressViewHolder pvh = ((ProgressViewHolder) holder);
			if (!value) {
				pvh.pbFooterProgress.setVisibility(View.VISIBLE);
				pvh.pbFooterProgress.setIndeterminate(true);
			} else pvh.pbFooterProgress.setVisibility(View.GONE);
		} else {
			final SimpleViewHolder vh = (SimpleViewHolder) holder;
			vh.setPrice(productList);
			vh.setProductName(productList);
			vh.setCartImage(productList);
			vh.setChecked(productList);
			vh.setDefaultQuantity();
			vh.showProgressBar(productList.viewIsLoading);
			vh.disableSwipeToDelete(false);
			vh.setTvColorSize(productList);
			vh.hideDropdownIcon();
			vh.cbxItem.setOnClickListener(new View.OnClickListener() {
				@Override
				public void onClick(View view) {
					/**
					 * Disable clothing type selection when product detail api is loading
					 * food item type can still be selected.
					 */
					ProductList productList = mProductList.get(vh.getAdapterPosition());
					String productType = productList.productType;
					if (!productType.equalsIgnoreCase(FOOD_PRODUCT)) {
						boolean unlockSelection = !viewIsLoading();
						vh.cbxItem.setChecked(unlockSelection);
						if (unlockSelection) {
							onCheckItemClick(vh);
						}
					} else {
						onCheckItemClick(vh);
					}
				}
			});

			vh.llItemContainer.setOnClickListener(new View.OnClickListener()

			{
				@Override
				public void onClick(View v) {
					if (!viewIsLoading()) {
						onItemClick(vh);
					}
				}
			});

			mItemManger.bindView(holder.itemView, position);
		}
	}

	private void onCheckItemClick(SimpleViewHolder vh) {
		int position = vh.getAdapterPosition();
		ProductList selectedProduct = mProductList.get(position);
		int otherSkuSize = getOtherSkuSize(selectedProduct);
		// ProductDetails of type clothing or OtherSkus > 0
		if (clothingTypeProduct(selectedProduct)) {
			selectedProduct.viewIsLoading = !selectedProduct.viewIsLoading;
			if (selectedProduct.itemWasChecked) selectedProduct.viewIsLoading = false;
			selectedProduct.itemWasChecked = productWasChecked(selectedProduct);
			mSearchResultNavigator.onCheckedItem(mProductList, selectedProduct, selectedProduct.viewIsLoading);
			notifyItemChanged(position);
		} else {
			selectedProduct.itemWasChecked = productWasChecked(selectedProduct);
			mSearchResultNavigator.onFoodTypeChecked(mProductList, selectedProduct);
			mSearchResultNavigator.minOneItemSelected(mProductList);
			notifyItemChanged(position);
		}
	}

	private boolean clothingTypeProduct(ProductList selectedProduct) {
		return !selectedProduct.productType.equalsIgnoreCase(FOOD_PRODUCT);
	}

	private void onItemClick(SimpleViewHolder vh) {
		int position = vh.getAdapterPosition();
		ProductList selectedProduct = mProductList.get(position);
		int otherSkuSize = getOtherSkuSize(selectedProduct);
		// ProductDetails of type clothing or OtherSkus > 0
		if (clothingTypeProduct(selectedProduct)) {
			mSearchResultNavigator.onClothingTypeSelect(selectedProduct);
		} else {
			mSearchResultNavigator.onFoodTypeSelect(selectedProduct);
		}
	}

	private int getOtherSkuSize(ProductList selectedProduct) {
		List<OtherSkus> otherSkuList = selectedProduct.otherSkus;
		int otherSkuSize = 0;
		if (otherSkuList != null) {
			otherSkuSize = otherSkuList.size();
		}
		return otherSkuSize;
	}

	private boolean productWasChecked(ProductList prodList) {
		return !prodList.itemWasChecked;
	}

	@Override
	public int getSwipeLayoutResourceId(int position) {
		return R.id.swipe;
	}


	private class SimpleViewHolder extends RecyclerView.ViewHolder {

		private WTextView tvTitle;
		private TextView tvPrice;
		private TextView tvWasPrice;
		private WTextView tvSaveText;
		private WTextView tvQuantity;
		private WTextView tvColorSize;
		private WrapContentDraweeView cartProductImage;
		private LinearLayout llItemContainer;
		private CheckBox cbxItem;
		private ProgressBar pbLoadProduct;
		private SwipeLayout swipeLayout;
		private ImageView imPrice;

		private SimpleViewHolder(View view) {
			super(view);
			tvTitle = view.findViewById(R.id.tvTitle);
			tvPrice = view.findViewById(R.id.tvPrice);
			tvWasPrice = view.findViewById(R.id.tvWasPrice);
			tvSaveText = view.findViewById(R.id.tvSaveText);
			tvColorSize = view.findViewById(R.id.tvColorSize);
			cartProductImage = view.findViewById(R.id.cartProductImage);
			llItemContainer = view.findViewById(R.id.llItemContainer);
			cbxItem = view.findViewById(R.id.btnDeleteRow);
			pbLoadProduct = view.findViewById(R.id.pbLoadProduct);
			tvQuantity = view.findViewById(R.id.tvQuantity);
			swipeLayout = view.findViewById(R.id.swipe);
			imPrice = view.findViewById(R.id.imPrice);
		}

		private void setDefaultQuantity() {
			tvQuantity.setText("1");
		}

		private void setCartImage(ProductList productItem) {
			cartProductImage.setImageURI(productItem.externalImageRef + ((productItem.externalImageRef.indexOf("?") > 0) ? "w=" + 85 + "&q=" + 85 : "?w=" + 85 + "&q=" + 85));
		}

		public void setProductName(ProductList productItem) {
			tvTitle.setText(Html.fromHtml(productItem.productName));
		}

		private void setPrice(ProductList productItem) {
			SearchResultPriceItem priceItem =new  SearchResultPriceItem();
			priceItem.setPrice(productItem, itemView, true);
		}

		private void setSaveText(ProductList productItem) {
			tvSaveText.setVisibility(TextUtils.isEmpty(productItem.saveText) ? View.GONE : View.VISIBLE);
			tvSaveText.setText(!TextUtils.isEmpty(productItem.saveText) ? productItem.saveText : "");
		}

		public void setChecked(ProductList productList) {
			cbxItem.setChecked(productList.itemWasChecked);
		}

		public void showProgressBar(boolean visible) {
			pbLoadProduct.setVisibility(visible ? View.VISIBLE : View.GONE);
			cbxItem.setVisibility(visible ? View.GONE : View.VISIBLE);
		}

		public void disableSwipeToDelete(boolean enable) {
			swipeLayout.setRightSwipeEnabled(enable);
		}

		public void setTvColorSize(ProductList productlist) {
			tvColorSize.setText(TextUtils.isEmpty(productlist.displayColorSizeText) ? "" : productlist.displayColorSizeText);
		}

		public void hideDropdownIcon() {
			imPrice.setVisibility(View.GONE);
		}
	}

	private class ProgressViewHolder extends RecyclerView.ViewHolder {
		private ProgressBar pbFooterProgress;

		private ProgressViewHolder(View v) {
			super(v);
			pbFooterProgress = v.findViewById(R.id.pbFooterProgress);
		}
	}

	private class HeaderViewHolder extends RecyclerView.ViewHolder {
		private TextView tvNumberOfItem;

		private HeaderViewHolder(View v) {
			super(v);
			tvNumberOfItem = v.findViewById(R.id.tvNumberOfItem);
		}

		private void setTotalItem(ProductList productList) {
			if (productList.numberOfItems != null)
				tvNumberOfItem.setText(String.valueOf(productList.numberOfItems));
		}
	}

	@Override
	public int getItemViewType(int position) {
		return mProductList.get(position).rowType.ordinal();
	}

	public void refreshAdapter(boolean value, List<ProductList> tempProductList) {
		this.value = value;
		this.mProductList = tempProductList;
		notifyDataSetChanged();
	}

	public void setCheckedProgressBar(ProductList productList) {
		if (mProductList != null) {
			for (ProductList pList : mProductList) {
				if (pList == productList) {
					pList.viewIsLoading = !pList.viewIsLoading;
				}
			}
			notifyDataSetChanged();
		}
	}

	public void setSelectedSku(ProductList selectedProduct, OtherSkus selectedSKU) {
		if (mProductList != null) {
			for (ProductList pList : mProductList) {
				if (pList == selectedProduct) {
					pList.sku = selectedSKU.sku;
					String colour = TextUtils.isEmpty(selectedSKU.colour) ? "" : selectedSKU.colour;
					String size = TextUtils.isEmpty(selectedSKU.size) ? "" : selectedSKU.size;
					boolean colourSize = TextUtils.isEmpty(colour) || TextUtils.isEmpty(size);
					pList.displayColorSizeText = colourSize ? (colour + "" + size) : (colour + ", " + size);
				}
			}
			notifyDataSetChanged();
		}
	}

	public void onDeselectSKU(ProductList selectedProduct, OtherSkus selectedSKU) {
		if (mProductList != null) {
			for (ProductList pList : mProductList) {
				if (pList == selectedProduct) {
					pList.itemWasChecked = false;
					pList.viewIsLoading = false;
					pList.displayColorSizeText = "";
					if (mSearchResultNavigator != null)
						mSearchResultNavigator.minOneItemSelected(mProductList);
					notifyDataSetChanged();
					return;
				}
			}
		}
	}

	@Override
	public int getItemCount() {
		return mProductList == null ? 0 : mProductList.size();
	}

	public boolean viewIsLoading() {
		if (mProductList != null) {
			for (ProductList pList : mProductList) {
				if (pList.viewIsLoading) {
					return true;
				}
			}
		}
		return false;
	}

	public List<ProductList> getProductList() {
		return mProductList;
	}
}

File: app/src/main/java/za/co/woolworths/financial/services/android/ui/adapters/SearchResultShopAdapter.kt
package za.co.woolworths.financial.services.android.ui.adapters

import android.content.Context
import android.text.Html
import android.view.LayoutInflater
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.awfs.coordination.R
import com.awfs.coordination.databinding.BottomProgressBarBinding
import com.awfs.coordination.databinding.ItemFoundLayoutBinding
import com.awfs.coordination.databinding.LayoutCartListProductItemBinding
import com.daimajia.swipe.adapters.RecyclerSwipeAdapter
import za.co.woolworths.financial.services.android.models.dto.OtherSkus
import za.co.woolworths.financial.services.android.models.dto.ProductList
import za.co.woolworths.financial.services.android.ui.adapters.holder.ProductListingViewType
import za.co.woolworths.financial.services.android.ui.adapters.holder.SearchResultPriceItem
import za.co.woolworths.financial.services.android.ui.fragments.shoppinglist.search.SearchResultNavigator
import za.co.woolworths.financial.services.android.util.AppConstant.Companion.CONST_NO_SIZE

class SearchResultShopAdapter(
    private val context: Context, var productList: List<ProductList>?,
    private val mSearchResultNavigator: SearchResultNavigator?
) : RecyclerSwipeAdapter<RecyclerView.ViewHolder>() {

    private var value = false

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (ProductListingViewType.values()[viewType]) {
            ProductListingViewType.HEADER -> getHeaderViewHolder(parent)
            ProductListingViewType.FOOTER -> getProgressViewHolder(parent)
            ProductListingViewType.PRODUCT -> getSimpleViewHolder(parent)
        }
    }

    private fun getProgressViewHolder(parent: ViewGroup): ProgressViewHolder {
        return ProgressViewHolder(
            BottomProgressBarBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        )
    }

    private fun getSimpleViewHolder(parent: ViewGroup): SimpleViewHolder {
        return SimpleViewHolder(
            LayoutCartListProductItemBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            )
        )
    }

    private fun getHeaderViewHolder(parent: ViewGroup): HeaderViewHolder {
        return HeaderViewHolder(
            ItemFoundLayoutBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        )
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val productList = productList!![position]
        when (holder) {
            is HeaderViewHolder -> {
                holder.setTotalItem(productList)
            }
            is ProgressViewHolder -> {
                if (!value) {
                    holder.itemBinding.pbFooterProgress.visibility = VISIBLE
                    holder.itemBinding.pbFooterProgress.isIndeterminate = true
                } else holder.itemBinding.pbFooterProgress.visibility = GONE
            }
            is SimpleViewHolder -> {
                holder.setPrice(productList)
                holder.itemBinding?.apply {
                    promotionalTextLayout.visibility = GONE
                    llQuantity.visibility = GONE
                    strikeThroughGroup.visibility = VISIBLE
                    promotionalTextLayout.visibility = GONE

                    cbShoppingList.visibility = VISIBLE
                    cbShoppingList.isChecked = productList.itemWasChecked
                    tvTitle.text = Html.fromHtml(productList.productName)
                    tvQuantity.setText("1")
                    tvColorSize.setText(productList.displayColorSizeText ?: "")
                    swipe.isRightSwipeEnabled = false
                }
                holder.setCartImage(productList)
                holder.showProgressBar(productList.viewIsLoading)
                holder.itemBinding.cbShoppingList.setOnClickListener {
                    /**
                     * Disable clothing type selection when product detail api is loading
                     * food item type can still be selected.
                     */
                    /**
                     * Disable clothing type selection when product detail api is loading
                     * food item type can still be selected.
                     */
                    val productList: ProductList? =
                        this@SearchResultShopAdapter.productList?.getOrNull(holder.adapterPosition)
                            ?: null
                    val productType = productList?.productType ?: ""
                    if (!productType.equals(FOOD_PRODUCT, ignoreCase = true)) {
                        val unlockSelection = !viewIsLoading()
                        holder.itemBinding.cbShoppingList.isChecked = unlockSelection
                        if (unlockSelection) {
                            onCheckItemClick(holder)
                        }
                    } else {
                        onCheckItemClick(holder)
                    }
                }
                holder.itemBinding.llItemContainer.setOnClickListener {
                    if (!viewIsLoading()) {
                        onItemClick(holder)
                    }
                }
                mItemManger.bindView(holder.itemBinding.root, position)
            }
        }
    }

    private fun onCheckItemClick(vh: SimpleViewHolder) {
        val position = vh.adapterPosition
        val selectedProduct = productList!![position]
        // ProductDetails of type clothing or OtherSkus > 0
        if (clothingTypeProduct(selectedProduct)) {
            selectedProduct.viewIsLoading = !selectedProduct.viewIsLoading
            if (selectedProduct.itemWasChecked) selectedProduct.viewIsLoading = false
            mSearchResultNavigator?.onCheckedItem(
                productList,
                selectedProduct,
                selectedProduct.viewIsLoading
            )
            notifyItemChanged(position)
        } else {
            selectedProduct.itemWasChecked = productWasChecked(selectedProduct)
            mSearchResultNavigator?.onFoodTypeChecked(productList, selectedProduct)
            mSearchResultNavigator?.minOneItemSelected(productList)
            notifyItemChanged(position)
        }
    }

    private fun clothingTypeProduct(selectedProduct: ProductList): Boolean {
        return !selectedProduct.productType.equals(FOOD_PRODUCT, ignoreCase = true)
    }

    private fun onItemClick(vh: SimpleViewHolder) {
        val position = vh.adapterPosition
        val selectedProduct = productList!![position]
        // ProductDetails of type clothing or OtherSkus > 0
        if (clothingTypeProduct(selectedProduct)) {
            mSearchResultNavigator!!.onClothingTypeSelect(selectedProduct)
        } else {
            mSearchResultNavigator!!.onFoodTypeSelect(selectedProduct)
        }
    }

    private fun productWasChecked(prodList: ProductList): Boolean {
        return !prodList.itemWasChecked
    }

    override fun getSwipeLayoutResourceId(position: Int): Int {
        return R.id.swipe
    }

    private inner class SimpleViewHolder(val itemBinding: LayoutCartListProductItemBinding) :
        RecyclerView.ViewHolder(itemBinding.root) {

        fun setCartImage(productItem: ProductList) {
            productItem.externalImageRefV2?.let { url ->
                itemBinding.cartProductImage?.setImageURI(
                    url + (if ((url.indexOf("?") > 0)) "w=" + 85 + "&q=" + 85 else "?w=" + 85 + "&q=" + 85)
                )
            }
        }

        fun setPrice(productItem: ProductList) {
            val priceItem = SearchResultPriceItem()
            priceItem.setPrice(productItem, itemBinding, true)
        }

        fun showProgressBar(visible: Boolean) {
            itemBinding.pbLoadProduct.visibility = if (visible) VISIBLE else GONE
            itemBinding.btnDeleteRow.visibility = if (visible) GONE else VISIBLE
        }
    }

    private inner class ProgressViewHolder(val itemBinding: BottomProgressBarBinding) :
        RecyclerView.ViewHolder(itemBinding.root) {

    }

    private inner class HeaderViewHolder(val itemBinding: ItemFoundLayoutBinding) :
        RecyclerView.ViewHolder(itemBinding.root) {

        fun setTotalItem(productList: ProductList) {
            if (null != productList.numberOfItems) itemBinding.tvNumberOfItem.text =
                productList.numberOfItems.toString()
            if (productList.numberOfItems == 1) {
                itemBinding.tvFoundItem.text = context.getString(R.string.product_item)
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return productList!![position].rowType.ordinal
    }

    fun refreshAdapter(value: Boolean, tempProductList: List<ProductList>?) {
        this.value = value
        productList = tempProductList
        notifyDataSetChanged()
    }

    fun setCheckedProgressBar(productList: ProductList) {
        if (this.productList != null) {
            for (pList: ProductList in this.productList!!) {
                if (pList === productList) {
                    pList.viewIsLoading = !pList.viewIsLoading
                }
            }
            notifyDataSetChanged()
        }
    }

    fun setSelectedSku(selectedProduct: ProductList, selectedSKU: OtherSkus) {

        productList?.forEachIndexed { index, productList ->
            if (productList === selectedProduct) {
                productList.sku = selectedSKU.sku
                productList.displayColorSizeText = buildColorSizeText(
                    selectedSKU.colour,
                    selectedSKU.size
                )
                notifyItemChanged(index, productList)
                return@forEachIndexed
            }
        }
    }

    private fun buildColorSizeText(colour: String?, size: String?): String = buildString {
        append(colour ?: "")
        if (!colour.isNullOrEmpty() && colour.equals(size, ignoreCase = true)) {
            return@buildString
        }

        if(size.isNullOrEmpty() || CONST_NO_SIZE.equals(size, ignoreCase = true)) {
            return@buildString
        }

        append(if (!colour.isNullOrEmpty()) ", " else "")
        append(size)
    }

    fun onDeselectSKU(selectedProduct: ProductList, selectedSKU: OtherSkus?) {
        if (productList != null) {
            for (pList: ProductList in productList!!) {
                if (pList === selectedProduct) {
                    pList.itemWasChecked = false
                    pList.viewIsLoading = false
                    pList.displayColorSizeText = ""
                    mSearchResultNavigator?.minOneItemSelected(
                        productList
                    )
                    notifyDataSetChanged()
                    return
                }
            }
        }
    }

    override fun getItemCount(): Int {
        return if (productList == null) 0 else productList!!.size
    }

    fun viewIsLoading(): Boolean {
        if (productList != null) {
            for (pList: ProductList in productList!!) {
                if (pList.viewIsLoading) {
                    return true
                }
            }
        }
        return false
    }

    fun setSelectedProduct(selectedProduct: ProductList) {
        // Ensure that the product list is not null
        if(productList.isNullOrEmpty()) {
            return
        }

        // Find the index of the selected product within the list
        val index = productList!!.indexOf(selectedProduct)
        if (index < 0 || index >= productList!!.size) {
            // Return if the index is out of range
            return
        }

        // Update the selected item's properties
        val selectedItem = productList!!.getOrNull(index)?.apply {
            viewIsLoading = false
            itemWasChecked = true
        }

        // Notify the adapter that the selected item has changed
        notifyItemChanged(index, selectedItem)
    }

    companion object {
        private val FOOD_PRODUCT = "foodProducts"
    }
}